File: src/main/java/org/zkoss/gridster/GridItem.java
package org.zkoss.gridster;

import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.impl.XulElement;

import java.io.IOException;

/**
 * An item in a {@link Gridster}.
 *
 * @author Sean Connolly
 */
public class GridItem extends XulElement implements Comparable<GridItem> {

    private Integer row = null;
    private Integer col = null;
    private Integer sizeX = null;
    private Integer sizeY = null;
    private Integer maxSizeX = null;
    private Integer maxSizeY = null;
    private Integer minSizeX = null;
    private Integer minSizeY = null;

    public Integer getRow() {
        return row;
    }

    public Integer getCol() {
        return col;
    }

    public Integer getSizeX() {
        return sizeX;
    }

    public Integer getSizeY() {
        return sizeY;
    }

    public Integer getMaxSizeX() {
        return maxSizeX;
    }

    public Integer getMaxSizeY() {
        return maxSizeY;
    }

    public Integer getMinSizeX() {
        return minSizeX;
    }

    public Integer getMinSizeY() {
        return minSizeY;
    }

    public void setRow(int row) {
        if (this.row == null || this.row != row) {
            this.row = row;
            smartUpdate("row", row);
        }
    }

    public void setCol(int col) {
        if (this.col == null || this.col != col) {
            this.col = col;
            smartUpdate("col", col);
        }
    }

    public void setSizeX(int sizeX) {
        if (this.sizeX == null || this.sizeX != sizeX) {
            this.sizeX = sizeX;
            smartUpdate("sizeX", sizeX);
        }
    }

    public void setSizeY(int sizeY) {
        if (this.sizeY == null || this.sizeY != sizeY) {
            this.sizeY = sizeY;
            smartUpdate("sizeY", sizeY);
        }
    }

    public void setMaxSizeX(int maxSizeX) {
        if (this.maxSizeX == null || this.maxSizeX != maxSizeX) {
            this.maxSizeX = maxSizeX;
            smartUpdate("maxSizeX", maxSizeX);
        }
    }

    public void setMaxSizeY(int maxSizeY) {
        if (this.maxSizeY == null || this.maxSizeY != maxSizeY) {
            this.maxSizeY = maxSizeY;
            smartUpdate("maxSizeY", maxSizeY);
        }
    }

    public void setMinSizeX(int minSizeX) {
        if (this.minSizeX == null || this.minSizeX != minSizeX) {
            this.minSizeX = minSizeX;
            smartUpdate("minSizeX", minSizeX);
        }
    }

    public void setMinSizeY(int minSizeY) {
        if (this.minSizeY == null || this.minSizeY != minSizeY) {
            this.minSizeY = minSizeY;
            smartUpdate("minSizeY", minSizeY);
        }
    }

    @Override
    protected void renderProperties(ContentRenderer renderer) throws IOException {
        super.renderProperties(renderer);
        render(renderer, "row", row);
        render(renderer, "col", col);
        render(renderer, "sizeX", sizeX);
        render(renderer, "sizeY", sizeY);
        render(renderer, "maxSizeX", maxSizeX);
        render(renderer, "maxSizeY", maxSizeY);
        render(renderer, "minSizeX", minSizeX);
        render(renderer, "minSizeY", minSizeY);
    }

    @Override
    public int compareTo(GridItem that) {
        if (this.row == null && that.row == null) {
            return 0;
        }
        if (this.row == null || that.row == null) {
            return (this.row == null) ? -1 : 1;
        }
        int order = Integer.compare(this.row, that.row);
        if (order != 0) {
            return order;
        }
        if (this.col == null && that.col == null) {
            return 0;
        }
        if (this.col == null || that.col == null) {
            return (this.col == null) ? -1 : 1;
        }
        return Integer.compare(this.col, that.col);
    }

}

File: src/main/resources/web/js/zul/gridster/GridItem.js
zul.gridster.GridItem = zk.$extends(zk.Widget, {
    $define: {
        row: function (r) {
            var n = this.$n();
            if (n)
                n.row = r;
        },
        col: function (c) {
            var n = this.$n();
            if (n)
                n.col = c;
        },
        sizeX: function (x) {
            var n = this.$n();
            if (n)
                n.sizeX = x;
        },
        sizeY: function (y) {
            var n = this.$n();
            if (n)
                n.sizeY = y;
        },
        maxSizeX: function (x) {
            var n = this.$n();
            if (n)
                n.maxSizeX = x;
        },
        maxSizeY: function (y) {
            var n = this.$n();
            if (n)
                n.maxSizeY = y;
        },
        minSizeX: function (x) {
            var n = this.$n();
            if (n)
                n.minSizeX = x;
        },
        minSizeY: function (y) {
            var n = this.$n();
            if (n)
                n.minSizeY = y;
        }
    },
    //super//
    domAttrs_: function (no) {
        var attr = this.$supers('domAttrs_', arguments),
            row = this._row && 'data-row="' + this._row + '"',
            col = this._col && 'data-col="' + this._col + '"',
            sizeX = this._sizeX && 'data-sizex="' + this._sizeX + '"',
            sizeY = this._sizeX && 'data-sizey="' + this._sizeY + '"',
            maxSizeX = this._maxSizeX && 'data-max-sizex="' + this._maxSizeX + '"',
            maxSizeY = this._maxSizeY && 'data-max-sizey="' + this._maxSizeY + '"',
            minSizeX = this._minSizeX && 'data-min-sizex="' + this._minSizeX + '"',
            minSizeY = this._minSizeY && 'data-min-sizey="' + this._minSizeY + '"';
        return [attr, row, col, sizeX, sizeY, maxSizeX, maxSizeY, minSizeX, minSizeY].filter(function (val) {
            return val;
        }).join(' ');
    }
});